File: week1/Programming-Assignment-1/stack_with_max/stack_with_max.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <algorithm>
#include <stack>

using std::cin;
using std::string;
using std::vector;
using std::cout;
using std::max_element;

class StackWithMax {
    std::stack<int> stack;
    std::stack<int> maxInStack;

public:

    void Push(int value) {
        stack.push(value);
        // check if it is the first element
        if (stack.size() == 1) {
            maxInStack.push(value);
            return;
        }

        if (value > maxInStack.top()) {
            maxInStack.push(value);
        } else {
            maxInStack.push(maxInStack.top());
        }
    }

    void Pop() {
        assert(stack.size());
        stack.pop();
        maxInStack.pop();
    }

    int Max() const {
        assert(stack.size());
        return maxInStack.top();
    }
};

int main() {
    int num_queries = 0;
    cin >> num_queries;

    string query;
    string value;

    StackWithMax stack;

    for (int i = 0; i < num_queries; ++i) {
        cin >> query;
        if (query == "push") {
            cin >> value;
            stack.Push(std::stoi(value));
        }
        else if (query == "pop") {
            stack.Pop();
        }
        else if (query == "max") {
            cout << stack.Max() << "\n";
        }
        else {
            assert(0);
        }
    }
    return 0;
}

File: week1/Programming-Assignment-1/stack_with_max/stackwithmax.py
#python3
import sys


class StackWithMax:
    def __init__(self):
        self.__stack = []
        self.__stack_max = []

    def push(self, a):
        self.__stack.append(a)
        # if __stack_max is empty, just add this value
        if len(self.__stack) == 1:
            self.__stack_max.append(a)
            return
        # add this value to __stack_max if it is a new max, else add old max
        if a > self.__stack_max[-1]:
            self.__stack_max.append(a)
        else:
            self.__stack_max.append(self.__stack_max[-1])

    def pop(self):
        assert(len(self.__stack))
        self.__stack.pop()
        self.__stack_max.pop()

    def max(self):
        assert(len(self.__stack))
        return self.__stack_max[-1]


if __name__ == '__main__':
    stack = StackWithMax()

    num_queries = int(sys.stdin.readline())
    for _ in range(num_queries):
        query = sys.stdin.readline().split()

        if query[0] == "push":
            stack.push(int(query[1]))
        elif query[0] == "pop":
            stack.pop()
        elif query[0] == "max":
            print(stack.max())
        else:
            assert 0
